sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -543,14 +544,52 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
 
   DeviceImplPtr Dev =
       (MustBuildOnSubdevice == PI_TRUE) ? DeviceImpl : RootDevImpl;
-  auto BuildF = [this, &M, &KSId, &ContextImpl, &Dev, Prg, &CompileOpts,
-                 &LinkOpts, &JITCompilationIsRequired, SpecConsts] {
-    auto Context = createSyclObjFromImpl<context>(ContextImpl);
-    auto Device = createSyclObjFromImpl<device>(Dev);
+  auto Context = createSyclObjFromImpl<context>(ContextImpl);
+  auto Device = createSyclObjFromImpl<device>(Dev);
+  const RTDeviceBinaryImage &Img =
+      getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+
+  // Check that device supports all aspects used by the kernel
+  const RTDeviceBinaryImage::PropertyRange &ARange =
+      Img.getDeviceRequirements();
+
+#define __SYCL_ASPECT(A, B, C)                                                 \
+  case aspect::A:                                                              \
+    return C;
+#define __SYCL_SPECIAL_ASPECT(A, B, C)
+  auto getAspectNameStr = [](aspect AspectNum) -> std::string {
+    switch (AspectNum) {
+#include <sycl/info/aspects.def>
+    default:
+      throw sycl::exception(
+          errc::kernel_not_supported,
+          "Unknown aspect " + std::to_string(static_cast<unsigned>(AspectNum)));
+    }
+  };
+#undef __SYCL_SPECIAL_ASPECT
+#undef __SYCL_ASPECT
 
-    const RTDeviceBinaryImage &Img =
-        getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+  for (RTDeviceBinaryImage::PropertyRange::ConstIterator It : ARange) {
+    auto KName = std::string((*It)->Name);
+    if (KName != "aspects")
+      continue;
+    ByteArray Aspects = DeviceBinaryProperty(*It).asByteArray();
+    // 8 because we need to skip 64-bits of size of the byte array
+    auto *AIt = reinterpret_cast<const std::uint32_t *>(&Aspects[8]);
```

Comment:
Why isn't this illegal type-punning?

Reply:
I see the same case in https://github.com/intel/llvm/blob/sycl/sycl/source/detail/program_manager/program_manager.cpp#L1241. Don't really get what you mean, could you please explain?

Reply:
https://stackoverflow.com/questions/63422076/why-is-type-punning-considered-ub

Reply:
I think we can do this because originally these aspects were uint32_t. Maybe @steffenlarsen  can add some clarity?

Reply:
In that case, I don't like how the interface is defined (i.e. we shouldn't be doing this magic *here*), but that would be outside this PR's scope.

Reply:
I have an old patch trying to amend the UB: https://github.com/intel/llvm/pull/5537. It is a little stale now though.

Reply:
I have updated https://github.com/intel/llvm/pull/5537 and added changes for this.

Reply:
I've uploaded an alternative approach in https://github.com/intel/llvm/pull/7023.